pkg/querier/block.go

```diff
@@ -80,91 +124,227 @@ func (b *BlockQuerier) Get(ctx context.Context, userID string, from, through mod
 			Value: m.Value,
 		})
 	}
+	return converted
+}
 
-	ctx = metadata.AppendToOutgoingContext(ctx, "user", userID)
-	seriesClient, err := client.Series(ctx, &storepb.SeriesRequest{
-		MinTime:                 int64(from),
-		MaxTime:                 int64(through),
-		Matchers:                converted,
+func (b *blocksQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
+	log, ctx := spanlogger.New(b.ctx, "blocksQuerier.LabelValues")
+	defer log.Span.Finish()
+
+	ctx = b.addUserToContext(ctx)
+	resp, err := b.client.LabelValues(ctx, &storepb.LabelValuesRequest{
+		Label:                   name,
+		PartialResponseDisabled: false,
 		PartialResponseStrategy: storepb.PartialResponseStrategy_ABORT,
 	})
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	return resp.Values, nil, nil
+}
+
+func (b *blocksQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	log, ctx := spanlogger.New(b.ctx, "blocksQuerier.LabelValues")
+	defer log.Span.Finish()
+
+	ctx = b.addUserToContext(ctx)
+	resp, err := b.client.LabelNames(ctx, &storepb.LabelNamesRequest{
+		PartialResponseDisabled: false,
+		PartialResponseStrategy: storepb.PartialResponseStrategy_ABORT,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return resp.Names, nil, nil
+}
+
+func (b *blocksQuerier) Close() error {
+	// nothing to do here.
+	return nil
+}
+
+// implementation of storage.SeriesSet, based streamed responses from store client.
+type blockQuerierSeriesSet struct {
+	ctx          context.Context
+	seriesClient storepb.Store_SeriesClient
+
+	err error
+
+	currSeries *storepb.Series
+	currChunks []storepb.AggrChunk
+
+	nextSeries *storepb.Series
+}
+
+func (bqss *blockQuerierSeriesSet) Next() bool {
```

Comment:
In general, I prefer to stream data between systems on demand, rather than fetching all data into memory first. gRPC is explicitly designed this way as well, so it just seemed like a good idea to use it. You're correct in that it adds some complexity. And this complexity would be unnecessary, if the data source already had all data available or in memory.

I assumed that client store is fetching data on demand, so streaming would make sense. But looking at the code (vendor/github.com/thanos-io/thanos/pkg/store/bucket.go:793), it seems that it fetches all data concurrently into memory, and only then it starts sending to the client. I'll check it in more detail.

Reply:
I also have mixed feelings whether it's better keeping the current implementation (in case in the future `BucketStore.Series()` will stream the response) or simplifying it based on the current use case. I'm usually more lean towards focusing on the current use case, but I'm open to discuss it.

Reply:
I will simplify it based on current use case.